src/vs/workbench/parts/terminal/electron-browser/terminalInstance.ts

```diff
@@ -772,6 +774,8 @@ export class TerminalInstance implements ITerminalInstance {
 		this._processManager.onProcessExit(exitCode => this._onProcessExit(exitCode));
 		this._processManager.onProcessData(data => this._onData.fire(data));
 
+		this._onProcessLaunching.fire(this);
```

Comment:
I think the ID may not be ready at this time (race condition), `this._processManager.onProcessReady` should instead be used as the new ID will definitely be set:

https://github.com/Microsoft/vscode/blob/cbd629998bdcc07bd4f98ecc3721b27a20317ec9/src/vs/workbench/parts/terminal/electron-browser/terminalProcessManager.ts#L130-L131

Reply:
Sorry, I don't get your point. I am not setting nor using the process id here. This onProcessLaunching event is fired just so that the terminal instance can create the promise and store the resolver which can be done as soon as a new process request is received. Or am I missing something?

Reply:
/cc @Tyriar

Reply:
+1 to using onProcessReady to ensure that the ID is already set.

Reply:
I'm obviously missing the point here. I mean, onProcessLaunching signals that a new process will be launched whereas onProcessIdReady signals that the process DID launch. They serve different purposes in my opinion.

Besides, why have these events nested? If onProcessLaunching is going to be fired by onProcessIdReady isn't it more sensible to just create the process Id promise in onProcessIdReady anyway?

Could you please elaborate on that just for the sake of clarity?

Reply:
You're using the event and then passing in `instance`

https://github.com/Microsoft/vscode/pull/60111/files#diff-738f70b1596a3687ce11b464cd29bfc2R283

That event listener then uses instance.id:

https://github.com/Microsoft/vscode/pull/60111/files#diff-fef4270273da47cf116c1d80fe3da649R38

This is what we're concerned about, `instance.id` may still be out of date but it wouldn't be if `this._processManager.onProcessReady` was used instead to fire the original event.

Reply:
@Tyriar Hmmm.. But isn't that the terminal instance id (rather than the process id) ? The terminal instance id is set in its constructor here:

https://github.com/Microsoft/vscode/blob/0ce96012870c5364eb143c62b9fcce56fcdfecf8/src/vs/workbench/parts/terminal/electron-browser/terminalInstance.ts#L142

And should not depend on the process id being set. Is that correct?